matchzoo/preprocessors.py

```diff
@@ -0,0 +1,133 @@
+"""Matchzoo toolkit for text pre-processing."""
+
+import re
+import nltk
+import jieba
+import many_stop_words
+from functools import reduce
+
+
+def get_stopwords(lang: str='en') -> list:
+    """
+    Get stopwords based on language.
+
+    :params lang: language code.
+
+    :return stop_list: list of stop words.
+    """
+    return many_stop_words.get_stop_words(lang)
+
+
+def tokenizer(raw: str) -> list:
+    """
+    Word tokenization.
+
+    :param raw: raw input given by user.
+
+    :return terms: list of tokenized terms.
+    """
+    return nltk.word_tokenize(raw)
+
+
+def segmentation(raw: str) -> list:
+    """
+    Chinese sentence segmentation.
+
+    :param raw: raw input given by user.
+
+    :return terms: list of segmentated terms.
+    """
+    return list(jieba.cut(raw))
+
```

Comment:
I don't think using two methods for two languages is a good idea. They're both essentially tokenizers. Use an optional language argument and decide what tokenizer backend to use on the fly.

Reply:
That's not really doable with the structure you have now since the `reduce` you're using there requires a uniform signature. Either decorate your functions with something like `functools.partial`, or make these function classes and use instances of these classes as pipeline units. I used `reduce` at first, then I switched to decorators, and finally I settled down on classes. Much better flexibility, and easier to serialise.

Reply:
only asian langauages such as chinese need segmentation..

Reply:
@uduse someting  like return function itself?:
```python```
class Preprocessor...

def func1(self, tokens):
    # do something
    return self
```

Reply:
Segmentation and tokenization are really the same thing. They're the processes of partitioning a sequence of characters into minimum processing units (terms). By using class, I mean making callable classes. 

e.g.

```python
class Lower:
    def __call__(self, tokens):
        return [token.lower() for token in tokens]
```